src/riak_kv_qry_compiler.erl

```diff
@@ -43,35 +45,61 @@
                         {start_inclusive, boolean()} |
                         {end_inclusive, boolean()}].
 -type combinator()       :: [binary()].
--type limit()            :: any().
+-type limit()            :: pos_integer().
+-type offset()           :: non_neg_integer().
 -type operator()         :: [binary()].
--type sorter()           :: term().
+-type sorter()           :: {binary(), asc|desc, nulls_first|nulls_last}.
 
 
 -export_type([
     combinator/0,
-    limit/0,
+    limit/0, offset/0,
     operator/0,
     sorter/0
 ]).
 -export_type([where_props/0]).
 
-%% 3rd argument is undefined if we should not be concerned about the
-%% maximum number of quanta
--spec compile(?DDL{}, ?SQL_SELECT{}, 'undefined'|pos_integer()) ->
+-spec compile(?DDL{}, ?SQL_SELECT{}) ->
     {ok, [?SQL_SELECT{}]} | {error, any()}.
-compile(?DDL{}, ?SQL_SELECT{is_executable = true}, _MaxSubQueries) ->
+compile(?DDL{}, ?SQL_SELECT{is_executable = true}) ->
     {error, 'query is already compiled'};
-compile(?DDL{ table = T} = DDL,
-        ?SQL_SELECT{is_executable = false} = Q1, MaxSubQueries) ->
+compile(?DDL{table = T} = DDL,
+        ?SQL_SELECT{is_executable = false} = Q1) ->
     Mod = riak_ql_ddl:make_module_name(T),
-    case maybe_compile_group_by(Mod, compile_select_clause(DDL, Q1), Q1) of
+    case compile_order_by(
+           maybe_compile_group_by(
+             Mod, compile_select_clause(DDL, Q1), Q1)) of
         {ok, Q2} ->
-            compile_where_clause(DDL, Q2, MaxSubQueries);
+            compile_where_clause(DDL, Q2);
         {error, _} = Error ->
             Error
     end.
 
+
+-spec compile_order_by({ok, ?SQL_SELECT{}} | {error, any()}) ->
+                              {ok, ?SQL_SELECT{}} | {error, any()}.
+compile_order_by({error,_} = E) ->
+    E;
+compile_order_by({ok, ?SQL_SELECT{'ORDER BY' = undefined} = Q}) ->
+    {ok, Q};
+compile_order_by({ok, ?SQL_SELECT{'ORDER BY' = OrderBy} = Q}) ->
+    case non_unique_identifiers(OrderBy) of
+        [] ->
+            {ok, Q?SQL_SELECT{'ORDER BY' = OrderBy}};
+        WhichNonUnique ->
+            {error, {non_unique_orderby_fields, ?E_NON_UNIQUE_ORDERBY_FIELDS(hd(WhichNonUnique))}}
+    end.
+
+non_unique_identifiers(FF) ->
+    Occurses = [{F, occurs(F, FF)} || {F, _, _} <- FF],
+    [F || {F, N} <- Occurses, N > 1].
```

Comment:
**Abstract**

English influences in modern Erlang are subtle and not well understood.

**Introduction**

`Occurses` may not be correct English but it *is* correct programmese.

**Discussion**

There is logic and consistency in calling `Poses` a list of positions, each denoted by variable `Pos` elsewhere in your code.

**Conclusions**

I beg to disagree.